sdk/src/nonce_instruction.rs

```diff
@@ -0,0 +1,479 @@
+use log::*;
+use num_derive::{FromPrimitive, ToPrimitive};
+use serde_derive::{Deserialize, Serialize};
+use solana_logger;
+use crate::{
+    account::{get_signers, KeyedAccount},
+    instruction::{AccountMeta, Instruction, InstructionError},
+    instruction_processor_utils::{limited_deserialize, next_keyed_account, DecodeError},
+    nonce_program::id,
+    nonce_state::{NonceAccount, NonceState},
+    pubkey::Pubkey,
+    system_instruction,
+    sysvar::{
+        Sysvar,
+        recent_blockhashes::{self, RecentBlockhashes},
+        rent::{self, Rent},
+    }
+};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, FromPrimitive, ToPrimitive)]
+pub enum NonceError {
+    NoRecentBlockhashes,
+    NotExpired,
+    UnexpectedValue,
+    BadAccountState,
+}
+
+impl<E> DecodeError<E> for NonceError {
+    fn type_of() -> &'static str {
+        "NonceError"
+    }
+}
+
+impl std::fmt::Display for NonceError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            NonceError::NoRecentBlockhashes => write!(f, "recent blockhash list is empty"),
+            NonceError::NotExpired => write!(f, "stored nonce is still in recent_blockhashes"),
+            NonceError::UnexpectedValue => write!(f, "specified nonce does not match stored nonce"),
+            NonceError::BadAccountState => write!(f, "cannot handle request in current account state"),
+        }
+    }
+}
+
+impl std::error::Error for NonceError {}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub enum NonceInstruction {
+    /// `Nonce` consumes a stored nonce, replacing it with a successor
+    ///
+    /// Expects 2 Accounts:
+    ///     0 - Initialized NonceAccount to be spent
+    ///     1 - RecentBlockhashes sysvar carrying a subset of the bank's
+    ///         BlockhashQueue
+    ///
+    /// The first `Hash` is the one to be spent, the second is its replacement
+    Nonce,
+
+    /// `Withdraw` transfers funds out of the nonce account
+    ///
+    /// Expects 3 Accounts:
+    ///     0 - A NonceAccount
+    ///     1 - A system account to which the lamports will be transferred
+    ///     2 - RecentBlockhashes sysvar carrying a subset of the bank's
+    ///         BlockhashQueue
+    ///
+    /// The `u64` parameter is the lamports to withdraw, which must leave the
+    /// account balance above the rent exempt reserve or at zero.
+    Withdraw(u64),
+}
+
+pub fn create_nonce_account(
```

Comment:
Maybe `create_nonce_account` should create a transaction that both creates the system account and initialized the nonce.  That way it would be atomic and we don't have the initial lamport balance separated from the min lamports required check during initialize

Reply:
There was originally a separate `Initialize` instruction to drive the account state from `Uninitialized` to `Initialized`.  Some of the tests that differentiated `Initialize` from `Nonce` had to be moved into runtime to prevent fee theft.  What was left didn't really warrant a separate instruction

Reply:
From a user's perspective, it would be nice if the API wrapped these into one atomic operation (create an account and initialize the nonce

Reply:
The `Initialize` instruction was issued atomically (WRT user API).  I can't think of any reason off the top of my head not to issue a `Nonce` similarly.  I'll review my convo with @rob-solana to be sure I'm not forgetting something and toss it in

Reply:
Was thinking something like this: https://github.com/solana-labs/solana/blob/c706f9b2cd194f44bffa78e64e84550a75211ab2/programs/vest/src/vest_instruction.rs#L87

Reply:
Yep!  It used to look like this:
https://github.com/solana-labs/solana/pull/6845/commits/33b9954e269f4e871f10a38444ca47573585b370#diff-37a2ca5217495a5ada65ad8207fcc32bL100-L110

Reply:
a separate Initialize instruction just seemed unnecessary.  Nonce does exactly one job, it always requires the Nonce account signature, and the actual instructions are hidden behind Vec<Instruction> library calls.

Reply:
@rob-solana, @jackcmay is asking if `create_nonce_account()` can return a `Nonce` instruction in addition to the `system_instruction::create_account`, as a convenience to SDK users.  I can't think of any reason not to

Reply:
yes, it can.